internal/state/store.go

```diff
@@ -407,60 +420,106 @@ func ABCIResponsesResultsHash(ar *tmstate.ABCIResponses) []byte {
 
+	if len(buf) == 0 {
 		return nil, ErrNoABCIResponsesForHeight{height}
 	}
-
 	abciResponses := new(tmstate.ABCIResponses)
 	err = abciResponses.Unmarshal(buf)
 	if err != nil {
 		// DATA HAS BEEN CORRUPTED OR THE SPEC HAS CHANGED
 		tmos.Exit(fmt.Sprintf(`LoadABCIResponses: Data has been corrupted or its spec has
-                changed: %v\n`, err))
+				changed: %v\n`, err))
 	}
 	// TODO: ensure that buf is completely read.
 
 	return abciResponses, nil
 }
 
+// LoadLastABCIResponses loads the last ABCI response specifically after a crash
+// We conduct a sanity check here which compares the height from the last ABCIResponse to the
+// height that is given
+//
+// This is used for recovering from crashes where we called app.Commit and
+// before we called s.Save()
+func (store dbStore) LoadLastABCIResponse(height int64) (*tmstate.ABCIResponses, error) {
+	bz, err := store.db.Get(lastABCIResponseKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(bz) == 0 {
+		return nil, errors.New("value retrieved from db is empty")
+	}
+
+	abciResponse := new(tmstate.ABCIResponsesInfo)
+	err = abciResponse.Unmarshal(bz)
+	if err != nil {
+		// DATA HAS BEEN CORRUPTED OR THE SPEC HAS CHANGED
+		tmos.Exit(fmt.Sprintf(`LoadLastABCIResponses: Data has been corrupted or its spec has
+			changed: %v\n`, err))
+	}
+
+	//Here we sanitize by comparing the height from the height that is given to the last ABCIResponse
+	if height != abciResponse.GetHeight() {
+		return nil, errors.New("expected height %d but last stored abci responses was at height %d")
+	}
+
+	return abciResponse.AbciResponses, nil
+}
+
 // SaveABCIResponses persists the ABCIResponses to the database.
 // This is useful in case we crash after app.Commit and before s.Save().
 // Responses are indexed by height so they can also be loaded later to produce
 // Merkle proofs.
 //
-// Exposed for testing.
+// CONTRACT: height must be monotonically increasing every time this is called
 func (store dbStore) SaveABCIResponses(height int64, abciResponses *tmstate.ABCIResponses) error {
-	return store.saveABCIResponses(height, abciResponses)
-}
-
-func (store dbStore) saveABCIResponses(height int64, abciResponses *tmstate.ABCIResponses) error {
 	var dtxs []*abci.ResponseDeliverTx
 	// strip nil values,
 	for _, tx := range abciResponses.DeliverTxs {
 		if tx != nil {
 			dtxs = append(dtxs, tx)
 		}
 	}
-
 	abciResponses.DeliverTxs = dtxs
 
-	bz, err := abciResponses.Marshal()
+	// If the flag is false then we save the ABCIResponse. This can be used for the /BlockResults
+	// query or to reindex an event using the command line.
+	if !store.discardABCIResponses {
+		bz, err := abciResponses.Marshal()
+		if err != nil {
+			return err
+		}
+		if err := store.db.Set(abciResponsesKey(height), bz); err != nil {
+			return err
+		}
+
+	}
+
+	// We always save the last ABCI response incase we crash after app.Commit and before s.Save(.)
+	// This overwrites the previous saved ABCI Response
```

Comment:
I'm not sure we need to be quite so specific in this comment?
```suggestion
	// We always save the last ABCI response for crash recovery.
	// This overwrites the previous saved ABCI Response.
```